Proteomics/Protein/Protein.cs

```diff
@@ -275,6 +278,73 @@ internal static IEnumerable<PeptideWithSetModifications> RemoveTerminalModificat
             }
         }
 
+        internal IEnumerable<PeptideWithSetModifications> GetCleavageModifiedPeptides(IEnumerable<PeptideWithSetModifications> originalPeptides, DigestionParams digestionParams)
+        {
+            if (digestionParams.Protease.CleavageMassShifts.Count ==0)
+            {
+                foreach(var peptide in originalPeptides)
+                {
+                    yield return peptide;    
+                }                
+            }
+            else
+            {
+                Dictionary<string, (int, double)> residueSpecifics = new Dictionary<string, (int, double)>();
+                var massShifts = digestionParams.Protease.CleavageMassShifts;
+                var digestionMotifs = digestionParams.Protease.DigestionMotifs;
+                foreach (var motif in digestionMotifs)
+                {
+                    var cleavageResidue = motif.InducingCleavage;
+                    // cutIndex of 0 is N and 1 is C
+                    var cleavageIndex = motif.CutIndex;
+                    if (massShifts.ContainsKey(cleavageResidue))
+                    {
+                        residueSpecifics.Add(cleavageResidue, (cleavageIndex, massShifts[cleavageResidue]));
+                    }
+                }
+
+                foreach (var peptide in originalPeptides)
+                {
+                    var tempPeptide = peptide;
+                    foreach (var modEvent in residueSpecifics)
+                    {
+                        var cleavageIndex = modEvent.Value.Item1;
+                        var massShift = modEvent.Value.Item2;
+                        var residue = modEvent.Key;
+
+                        if (cleavageIndex == 0) //look for residue at begining of peptide 
+                        {
+                            if (peptide.PreviousAminoAcid != '-')
+                            {
+                                var nTermRes = tempPeptide.BaseSequence.Substring(0, 1);
+                                if (residue == nTermRes)
+                                {
+                                    var newMass = tempPeptide.MonoisotopicMass + massShift;
+                                    tempPeptide.SetMonoisotopicMass(newMass);
+                                }
+                            }
+
+                        }
+                        else if (cleavageIndex == 1) //look for residue at end of peptide
+                        {
+                            if (peptide.NextAminoAcid != '-')
```

Comment:
use the protein.Length variable to determine the location in the protein sequence instead of '-'